File: android/src/main/java/com/mycompany/plugins/example/core/FileOperations.java
package com.mycompany.plugins.example.core;

import android.content.Context;
import android.util.Base64;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

/**
 * 文件操作模块
 * 负责文件的 CRUD 操作：创建、读取、写入、删除、重命名、移动、复制
 */
public class FileOperations {
    private static final String TAG = "FileOperations";
    private final Context context;

    public FileOperations(Context context) {
        this.context = context;
    }

    /**
     * 创建文件
     */
    public void createFile(String path, String content, String encoding) throws Exception {
        File file = new File(path);

        // 确保父目录存在
        File parentDir = file.getParentFile();
        if (parentDir != null && !parentDir.exists()) {
            if (!parentDir.mkdirs()) {
                throw new Exception("Failed to create parent directories for: " + path);
            }
        }

        try (FileOutputStream fos = new FileOutputStream(file)) {
            byte[] data;
            if ("base64".equals(encoding)) {
                data = Base64.decode(content, Base64.DEFAULT);
            } else {
                data = content.getBytes(StandardCharsets.UTF_8);
            }
            fos.write(data);
        } catch (IOException e) {
            throw new Exception("Failed to create file: " + e.getMessage());
        }
    }

    /**
     * 读取文件
     */
    public JSObject readFile(String path, String encoding) throws Exception {
        File file = new File(path);

        if (!file.exists()) {
            throw new Exception("File does not exist: " + path);
        }

        if (!file.isFile()) {
            throw new Exception("Path is not a file: " + path);
        }

        try (FileInputStream fis = new FileInputStream(file)) {
            byte[] data = new byte[(int) file.length()];
            fis.read(data);

            String content;
            if ("base64".equals(encoding)) {
                content = Base64.encodeToString(data, Base64.DEFAULT);
            } else {
                content = new String(data, StandardCharsets.UTF_8);
            }

            JSObject result = new JSObject();
            result.put("content", content);
            result.put("encoding", encoding);

            return result;
        } catch (IOException e) {
            throw new Exception("Failed to read file: " + e.getMessage());
        }
    }

    /**
     * 写入文件
     */
    public void writeFile(String path, String content, String encoding, boolean append) throws Exception {
        File file = new File(path);

        // 确保父目录存在
        File parentDir = file.getParentFile();
        if (parentDir != null && !parentDir.exists()) {
            if (!parentDir.mkdirs()) {
                throw new Exception("Failed to create parent directories for: " + path);
            }
        }

        try (FileOutputStream fos = new FileOutputStream(file, append)) {
            byte[] data;
            if ("base64".equals(encoding)) {
                data = Base64.decode(content, Base64.DEFAULT);
            } else {
                data = content.getBytes(StandardCharsets.UTF_8);
            }
            fos.write(data);
        } catch (IOException e) {
            throw new Exception("Failed to write file: " + e.getMessage());
        }
    }

    /**
     * 删除文件
     */
    public void deleteFile(String path) throws Exception {
        File file = new File(path);

        if (!file.exists()) {
            throw new Exception("File does not exist: " + path);
        }

        if (!file.isFile()) {
            throw new Exception("Path is not a file: " + path);
        }

        if (!file.delete()) {
            throw new Exception("Failed to delete file: " + path);
        }
    }

    /**
     * 重命名文件或目录
     */
    public void renameFile(String path, String newName) throws Exception {
        File file = new File(path);

        if (!file.exists()) {
            throw new Exception("File or directory does not exist: " + path);
        }

        File parentDir = file.getParentFile();
        if (parentDir == null) {
            throw new Exception("Cannot get parent directory for: " + path);
        }

        File newFile = new File(parentDir, newName);

        if (newFile.exists()) {
            throw new Exception("A file or directory with that name already exists: " + newFile.getPath());
        }

        if (!file.renameTo(newFile)) {
            throw new Exception("Failed to rename: " + path + " to " + newName);
        }
    }

    /**
     * 移动文件
     */
    public void moveFile(String sourcePath, String destinationPath) throws Exception {
        File sourceFile = new File(sourcePath);
        File destFile = new File(destinationPath);

        if (!sourceFile.exists()) {
            throw new Exception("Source file does not exist: " + sourcePath);
        }

        // 确保目标父目录存在
        File destParent = destFile.getParentFile();
        if (destParent != null && !destParent.exists()) {
            if (!destParent.mkdirs()) {
                throw new Exception("Failed to create destination directory");
            }
        }

        // 尝试直接重命名（同一文件系统内更快）
        if (!sourceFile.renameTo(destFile)) {
            // 如果重命名失败，尝试复制后删除
            copyFileInternal(sourceFile, destFile);
            if (!FileUtils.deleteRecursively(sourceFile)) {
                throw new Exception("Failed to delete source after copy");
            }
        }
    }

    /**
     * 复制文件
     */
    public void copyFile(String sourcePath, String destinationPath, boolean overwrite) throws Exception {
        File sourceFile = new File(sourcePath);
        File destFile = new File(destinationPath);

        if (!sourceFile.exists()) {
            throw new Exception("Source file does not exist: " + sourcePath);
        }

        if (destFile.exists() && !overwrite) {
            throw new Exception("Destination already exists: " + destinationPath);
        }

        // 确保目标父目录存在
        File destParent = destFile.getParentFile();
        if (destParent != null && !destParent.exists()) {
            if (!destParent.mkdirs()) {
                throw new Exception("Failed to create destination directory");
            }
        }

        copyFileInternal(sourceFile, destFile);
    }

    /**
     * 获取文件信息
     */
    public JSObject getFileInfo(String path) throws Exception {
        File file = new File(path);

        if (!file.exists()) {
            throw new Exception("File does not exist: " + path);
        }

        return FileUtils.createFileInfo(file);
    }

    /**
     * 检查文件是否存在
     */
    public boolean exists(String path) {
        File file = new File(path);
        return file.exists();
    }

    /**
     * 内部复制方法
     */
    private void copyFileInternal(File source, File dest) throws Exception {
        if (source.isDirectory()) {
            if (!dest.exists()) {
                dest.mkdirs();
            }
            File[] children = source.listFiles();
            if (children != null) {
                for (File child : children) {
                    copyFileInternal(child, new File(dest, child.getName()));
                }
            }
        } else {
            try (FileInputStream fis = new FileInputStream(source);
                 FileOutputStream fos = new FileOutputStream(dest)) {
                byte[] buffer = new byte[8192];
                int length;
                while ((length = fis.read(buffer)) > 0) {
                    fos.write(buffer, 0, length);
                }
            }
        }
    }
}

File: ios/Sources/AdvancedFileManagerPlugin/Core/FileOperations.swift
import Foundation

/// 文件操作模块
/// 负责文件的 CRUD 操作
public class FileOperations {
    private let fileManager = FileManager.default
    
    // MARK: - Create File
    
    public func createFile(path: String, content: String, encoding: String) throws {
        let url = URL(fileURLWithPath: path)
        
        // 确保父目录存在
        let parentDir = url.deletingLastPathComponent()
        if !fileManager.fileExists(atPath: parentDir.path) {
            try fileManager.createDirectory(at: parentDir, withIntermediateDirectories: true, attributes: nil)
        }
        
        let data: Data
        if encoding == "base64" {
            guard let decodedData = Data(base64Encoded: content) else {
                throw NSError(domain: "FileOperations", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 content"])
            }
            data = decodedData
        } else {
            guard let stringData = content.data(using: .utf8) else {
                throw NSError(domain: "FileOperations", code: 400, userInfo: [NSLocalizedDescriptionKey: "Failed to encode content as UTF-8"])
            }
            data = stringData
        }
        
        try data.write(to: url)
    }
    
    // MARK: - Read File
    
    public func readFile(path: String, encoding: String) throws -> [String: Any] {
        let url = URL(fileURLWithPath: path)
        
        guard fileManager.fileExists(atPath: path) else {
            throw NSError(domain: "FileOperations", code: 404, userInfo: [NSLocalizedDescriptionKey: "File does not exist: \(path)"])
        }
        
        var isDirectory: ObjCBool = false
        fileManager.fileExists(atPath: path, isDirectory: &isDirectory)
        
        guard !isDirectory.boolValue else {
            throw NSError(domain: "FileOperations", code: 400, userInfo: [NSLocalizedDescriptionKey: "Path is a directory: \(path)"])
        }
        
        let data = try Data(contentsOf: url)
        
        let content: String
        if encoding == "base64" {
            content = data.base64EncodedString()
        } else {
            guard let stringContent = String(data: data, encoding: .utf8) else {
                throw NSError(domain: "FileOperations", code: 500, userInfo: [NSLocalizedDescriptionKey: "Failed to decode file as UTF-8"])
            }
            content = stringContent
        }
        
        return [
            "content": content,
            "encoding": encoding
        ]
    }
    
    // MARK: - Write File
    
    public func writeFile(path: String, content: String, encoding: String, append: Bool) throws {
        let url = URL(fileURLWithPath: path)
        
        // 确保父目录存在
        let parentDir = url.deletingLastPathComponent()
        if !fileManager.fileExists(atPath: parentDir.path) {
            try fileManager.createDirectory(at: parentDir, withIntermediateDirectories: true, attributes: nil)
        }
        
        let data: Data
        if encoding == "base64" {
            guard let decodedData = Data(base64Encoded: content) else {
                throw NSError(domain: "FileOperations", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 content"])
            }
            data = decodedData
        } else {
            guard let stringData = content.data(using: .utf8) else {
                throw NSError(domain: "FileOperations", code: 400, userInfo: [NSLocalizedDescriptionKey: "Failed to encode content as UTF-8"])
            }
            data = stringData
        }
        
        if append && fileManager.fileExists(atPath: path) {
            let fileHandle = try FileHandle(forWritingTo: url)
            fileHandle.seekToEndOfFile()
            fileHandle.write(data)
            fileHandle.closeFile()
        } else {
            try data.write(to: url)
        }
    }
    
    // MARK: - Delete File
    
    public func deleteFile(path: String) throws {
        guard fileManager.fileExists(atPath: path) else {
            throw NSError(domain: "FileOperations", code: 404, userInfo: [NSLocalizedDescriptionKey: "File does not exist: \(path)"])
        }
        
        var isDirectory: ObjCBool = false
        fileManager.fileExists(atPath: path, isDirectory: &isDirectory)
        
        guard !isDirectory.boolValue else {
            throw NSError(domain: "FileOperations", code: 400, userInfo: [NSLocalizedDescriptionKey: "Path is a directory, use deleteDirectory: \(path)"])
        }
        
        try fileManager.removeItem(atPath: path)
    }
    
    // MARK: - Rename File
    
    public func renameFile(path: String, newName: String) throws {
        guard fileManager.fileExists(atPath: path) else {
            throw NSError(domain: "FileOperations", code: 404, userInfo: [NSLocalizedDescriptionKey: "File does not exist: \(path)"])
        }
        
        let url = URL(fileURLWithPath: path)
        let newURL = url.deletingLastPathComponent().appendingPathComponent(newName)
        
        guard !fileManager.fileExists(atPath: newURL.path) else {
            throw NSError(domain: "FileOperations", code: 409, userInfo: [NSLocalizedDescriptionKey: "A file with that name already exists: \(newURL.path)"])
        }
        
        try fileManager.moveItem(at: url, to: newURL)
    }
    
    // MARK: - Move File
    
    public func moveFile(sourcePath: String, destinationPath: String) throws {
        guard fileManager.fileExists(atPath: sourcePath) else {
            throw NSError(domain: "FileOperations", code: 404, userInfo: [NSLocalizedDescriptionKey: "Source file does not exist: \(sourcePath)"])
        }
        
        let destURL = URL(fileURLWithPath: destinationPath)
        let parentDir = destURL.deletingLastPathComponent()
        
        if !fileManager.fileExists(atPath: parentDir.path) {
            try fileManager.createDirectory(at: parentDir, withIntermediateDirectories: true, attributes: nil)
        }
        
        try fileManager.moveItem(atPath: sourcePath, toPath: destinationPath)
    }
    
    // MARK: - Copy File
    
    public func copyFile(sourcePath: String, destinationPath: String, overwrite: Bool) throws {
        guard fileManager.fileExists(atPath: sourcePath) else {
            throw NSError(domain: "FileOperations", code: 404, userInfo: [NSLocalizedDescriptionKey: "Source file does not exist: \(sourcePath)"])
        }
        
        let destURL = URL(fileURLWithPath: destinationPath)
        let parentDir = destURL.deletingLastPathComponent()
        
        if !fileManager.fileExists(atPath: parentDir.path) {
            try fileManager.createDirectory(at: parentDir, withIntermediateDirectories: true, attributes: nil)
        }
        
        if fileManager.fileExists(atPath: destinationPath) {
            if overwrite {
                try fileManager.removeItem(atPath: destinationPath)
            } else {
                throw NSError(domain: "FileOperations", code: 409, userInfo: [NSLocalizedDescriptionKey: "Destination already exists: \(destinationPath)"])
            }
        }
        
        try fileManager.copyItem(atPath: sourcePath, toPath: destinationPath)
    }
    
    // MARK: - Get File Info
    
    public func getFileInfo(path: String) throws -> [String: Any] {
        guard fileManager.fileExists(atPath: path) else {
            throw NSError(domain: "FileOperations", code: 404, userInfo: [NSLocalizedDescriptionKey: "File does not exist: \(path)"])
        }
        
        let url = URL(fileURLWithPath: path)
        let resourceValues = try url.resourceValues(forKeys: [
            .fileSizeKey,
            .contentModificationDateKey,
            .creationDateKey,
            .isDirectoryKey,
            .isHiddenKey
        ])
        
        return FileUtils.createFileInfo(from: url, resourceValues: resourceValues)
    }
    
    // MARK: - Exists
    
    public func exists(path: String) -> Bool {
        return fileManager.fileExists(atPath: path)
    }
}
